### PR TITLE
Optional LDAP integration.

### PR DESCRIPTION
**Pull request by caiush**
I changed the default (development) identify back end to SQL but enabled the ability to go against and existing ldap server. Should you wish to do such madness you need several things in your env file:

```
      "keystone" : {
          "backend" : "ldap" 
      },        
      "ldap" : {
         "config" : {
             "url" : "ldap://1.2.3.4",
             "user" : "CN=the_boss,OU=rock_stars,DC=example,DC=com",
             "password" : "sshhhh",
             "suffix" : "DC=example,DC=com",
             "use_dumb_member" :  "True",
             "query_scope" : "sub",
             "page_size" : "100",
             "user_tree_dn" : "OU=users,DC=example,DC=com",
             "user_objectclass" : "person",
             "user_id_attribute" : "cn",
             "user_name_attribute" : "name",
             "user_pass_attribute" : "",
             "user_allow_create" : "false",
             "user_allow_update" : "false",
             "user_allow_delete" : "false",
             "group_tree_dn" : "OU=groups,DC=example,DC=com",
             "group_objectclass" : "group",
             "group_allow_create" : "false",
             "group_allow_update" : "false",
             "group_allow_delete" : "false"
         },
         "admin_user" : "bcpc_admin",
         "admin_pass" : "super_secret"       
      },
```

The `admin_user` and `admin_pass` are the user/pass (already in ldap) you want to use as the cluster superuser. 

The PR assumes you are integrating against a corp LDAP/AD server to which you have no write privs. The onboard LDAP server is now not required and we should prob remove that. 

I also cleaned up some mess in the keystone recipes where a resource was doing 5 things in bash.

### CODE REVIEW

File: cookbooks/bcpc/recipes/keystone.rb [hunk 1]
```diff
@@ -156,42 +156,57 @@
...
+    user "root"
+    code <<-EOH
+        . /root/adminrc
+        . /root/keystonerc
+        keystone tenant-create --name "#{node['bcpc']['admin_tenant']}" --description "Admin services" 
+    EOH
+    not_if ". /root/keystonerc; . /root/adminrc; keystone tenant-get '#{node['bcpc']['admin_tenant']}'"
 end
 
-ruby_block "keystone-create-test-tenants" do
```
Comment (erhudy) [inline cookbooks/bcpc/recipes/keystone.rb]: Is removal of the test tenant intentional?

**Comment by erhudy** [comment]
I deployed this over top an existing local cluster where local 389DS was set up. I did not configure any of the LDAP stuff. It actually went pretty well, although since the process changes the UUIDs of existing users/tenancies it causes running instances to be orphaned from their tenancies, so doing this on a live cluster would be, unsurprisingly, not a good idea.

File: cookbooks/bcpc/recipes/keystone.rb (hunk 1 above)
Comment (caiush) [inline cookbooks/bcpc/recipes/keystone.rb]: aye

**Comment by caiush** [comment]
yes, sorry I should say *this is not chef-able into an existing cluster in any way*. There is zero migration path.

**Comment by erhudy** [comment]
I am re-testing everything on a fresh build.

**Comment by erhudy** [comment]
When generating the YAML checks configuration, it blows up now due to the removal of the test user:
```
  * template[/usr/local/etc/checks/default.yml] action create
    
    ================================================================================
    Error executing action `create` on resource 'template[/usr/local/etc/checks/default.yml]'
    ================================================================================
    
    Chef::Mixin::Template::TemplateError
    ------------------------------------
    No config found for get_config(keystone-test-user)!!!
    
    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/bcpc/recipes/checks-common.rb
    
     16:  template  "/usr/local/etc/checks/default.yml" do
     17:    source "checks/default.yml.erb"
     18:    owner "root"
     19:    group "root"
     20:    mode 00640
     21:  end
     22: 
    
    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/bcpc/recipes/checks-common.rb:16:in `from_file'
    
    template("/usr/local/etc/checks/default.yml") do
      action "create"
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      path "/usr/local/etc/checks/default.yml"
      backup 5
      atomic_update true
      source "checks/default.yml.erb"
      declared_type :template
      cookbook_name "bcpc"
      recipe_name "checks-common"
      mode 416
      owner "root"
      group "root"
    end
    
    Template Context:
    -----------------
    on line #4
      2: 
      3: tenant: <%=@node['bcpc']['admin_tenant']%>
      4: user: '<%="#{get_config('keystone-test-user')}"%>'
      5: password: '<%="#{get_config('keystone-test-password')}"%>'
      6: auth_url: https://<%="#{node[:bcpc][:management][:vip]}"%>:5000/v2.0/
```

**Comment by erhudy** [comment]
Fixed with @caiush's last push.

**Comment by kamidzi** [comment]
Could we excise b11f0e77646192e16dc9432a863f3b4aa854223d ?

**Comment by caiush** [comment]
nuked that commit

**Pull request merged**